### PR TITLE
ZK-3853: Tree supports 3-states selection

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zul/src/archive/web/js/zul/sel/less/tree.less [hunk 1]
```diff
@@ -282,4 +282,7 @@ body:not(.gecko) {
 	height: @meshAutoPagingRowHeight;
 	overflow: hidden;
 }
+.partial .z-treerow-checkbox{
+	background-color: #63B8E6;
```
Comment (DevChu) [inline zul/src/archive/web/js/zul/sel/less/tree.less]: the css class naming should follow the rule, like z-xxx-partial

File: zul/src/org/zkoss/zul/ext/TriStateTreeModel.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                
+        Description:
+                
+        History:
+                Thu Jun 07 16:07:43 CST 2018, Created by klyve
+
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+*/
+
+public interface TriStateTreeModel<E> {
```
Comment (DevChu) [inline zul/src/org/zkoss/zul/ext/TriStateTreeModel.java]: the javadoc should show more detail about this Model, please refer to TreeModel

File: zul/src/org/zkoss/zul/ext/TriStateTreeModel.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	 * @param selected
+	 */
+	public void toggleAncestors(TreeNode<E> node, boolean selected);
+
+	/**
+	 * A method to calculate the selection state of the tree
+	 * @param node
+	 * @return selection state
+	 */
+	public SelectionState calulateSelectionState(TreeNode<E> node);
```
Comment (DevChu) [inline zul/src/org/zkoss/zul/ext/TriStateTreeModel.java]: typo

File: zul/src/org/zkoss/zul/Treeitem.java [hunk 4]
```diff
@@ -673,6 +674,14 @@ protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer) th
 		if (_value instanceof String)
 			render(renderer, "value", _value);
 	}
+	
+	/*package*/ void setSelectionState(TriStateTreeModel.SelectionState state) {
+		if (state == TriStateTreeModel.SelectionState.PARTIAL) {
+			this.getTreerow().setSclass("z-treerow-partial");
+		} else {
+			this.getTreerow().setSclass("");
```
Comment (DevChu) [inline zul/src/org/zkoss/zul/Treeitem.java]: what if there are some customized sclass?

File: zul/src/org/zkoss/zul/Treeitem.java [hunk 5]
```diff
@@ -673,6 +674,14 @@ protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer) th
 		if (_value instanceof String)
 			render(renderer, "value", _value);
 	}
+	
+	/*package*/ void setSelectionState(TriStateTreeModel.SelectionState state) {
+		if (state == TriStateTreeModel.SelectionState.PARTIAL) {
+			this.getTreerow().smartUpdate("partialState",true);
+		} else {
```
Comment (DevChu) [inline zul/src/org/zkoss/zul/Treeitem.java]: code style
this.getTreerow().smartUpdate("partialState", true);

File: zul/src/archive/web/js/zul/sel/Treerow.js [hunk 6]
```diff
@@ -17,6 +17,14 @@ it will be useful, but WITHOUT ANY WARRANTY.
  * <p>Default {@link #getZclass}: z-treerow
  */
 zul.sel.Treerow = zk.$extends(zul.Widget, {
+	
+	setPartialState: function (partial) {
```
Comment (DevChu) [inline zul/src/archive/web/js/zul/sel/Treerow.js]: why don't just naming -> setSelectionState?

**Comment by scribetw** [comment]
06/29: postponed for next version

**Pull request closed without merging**